clients/src/main/java/org/apache/kafka/common/telemetry/internals/MetricNamingConvention.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.telemetry.internals;
+
+import org.apache.kafka.common.MetricName;
+
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * This class encapsulates naming and mapping conventions defined as part of
+ * <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Metricsnamingandformat">Metrics naming and format</a>
+ */
+public class MetricNamingConvention {
+
+    private static final String NAME_JOINER = ".";
+    private static final String TAG_JOINER = "_";
+
+    // remove metrics as it is redundant for telemetry metrics naming convention
+    private final static Pattern GROUP_PATTERN = Pattern.compile("\\.(metrics)");
+
+    public static MetricNamingStrategy<MetricName> getClientTelemetryMetricNamingStrategy(String domain) {
```

Comment:
If this is an internal class, do we expect to pass in anything else but `org.apache.kafka`?

Reply:
Maybe we can call it `prefix` instead of domain?

I believe this class should be reusable to also expose broker, streams, etc. metrics in OpenTelemetry format one day, so I would suggest to keep it.

Reply:
Renamed `domain` -> `prefix`.